streams/tck/src/main/java/org/eclipse/microprofile/reactive/streams/tck/OnStagesVerification.java

```diff
@@ -0,0 +1,232 @@
+    assertFalse(called.get());
+  }
+
+  @Test(expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = "failed")
+  public void onCompleteStageShouldPropagateRuntimeExceptions() {
+    await(ReactiveStreams.of("foo")
+        .onComplete(() -> {
+          throw new RuntimeException("failed");
+        })
+        .toList()
+        .run(getEngine()));
+  }
+
+  @Test
+  public void onCompleteStageShouldNotBeCalledWhenTheStreamFailed() {
+    AtomicBoolean called = new AtomicBoolean();
+    await(ReactiveStreams.failed(new Exception("failed"))
+      .onComplete(() -> called.set(true))
+      .toList()
+      .run(getEngine())
+      .exceptionally(t -> Collections.emptyList())
+    );
+    assertFalse(called.get());
+  }
+
+  @Test
+  public void onErrorStageShouldBeCalledWhenTheStreamFailed() {
+    AtomicReference<Throwable> called = new AtomicReference<>();
+    await(ReactiveStreams.failed(new Exception("failed"))
+      .onError(called::set)
+      .toList()
+      .run(getEngine())
+      .exceptionally(t -> Collections.emptyList())
+    );
+    assertNotNull(called.get());
+    assertTrue(called.get().getMessage().equalsIgnoreCase("failed"));
+  }
+
+  @Test
+  public void onTerminateStageShouldBeCalledWhenTheStreamFailed() {
+    AtomicBoolean called = new AtomicBoolean();
+    await(ReactiveStreams.failed(new Exception("failed"))
+      .onTerminate(() -> called.set(true))
+      .toList()
+      .run(getEngine())
+      .exceptionally(t -> Collections.emptyList())
+    );
+    assertTrue(called.get());
+  }
+
+  @Test(expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = "failed")
+  public void onTerminateStageShouldPropagateRuntimeExceptions() {
+    await(ReactiveStreams.of("foo")
+      .onTerminate(() -> {
+        throw new RuntimeException("failed");
+      })
+      .toList()
+      .run(getEngine()));
+  }
+
+  @Test
+  public void onCompleteAndOnTerminateStagesShouldBeCalledWhenTheStreamCompletes() {
+    AtomicBoolean onTerminateCalled = new AtomicBoolean();
+    AtomicBoolean onCompleteCalled = new AtomicBoolean();
+
+    await(ReactiveStreams.empty()
+      .onComplete(() -> onCompleteCalled.set(true))
+      .onTerminate(() -> onTerminateCalled.set(true))
+      .toList()
+      .run(getEngine())
+    );
+    assertTrue(onTerminateCalled.get());
+    assertTrue(onCompleteCalled.get());
+  }
+
+  @Test
+  public void onErrorAndOnTerminateStagesShouldBeCalledWhenTheStreamFails() {
+    AtomicBoolean onTerminateCalled = new AtomicBoolean();
+    AtomicReference<Throwable> onErrorCalled = new AtomicReference<>();
+
+    await(ReactiveStreams.failed(new Exception("failed"))
+      .onError(onErrorCalled::set)
+      .onTerminate(() -> onTerminateCalled.set(true))
+      .toList()
+      .run(getEngine())
+      .exceptionally(t -> Collections.emptyList())
+    );
+    assertTrue(onTerminateCalled.get());
+    assertEquals(onErrorCalled.get().getMessage(), "failed");
+  }
```

Comment:
Some additional tests:

I think we want to make sure `onTerminate` is called even when the stream is cancelled from downstream.

```java
CompletableFuture<Void> onTerminateCalled = new CompletableFuture<>();
ReactiveStreams.of(1)
  .flatMapCompletionStage(i -> new CompletableFuture())
  .onTerminate(() -> onTerminateCalled.complete(null))
  .cancel()
  .run(getEngine());
await(onTerminateCalled);
```

Also it's very likely that:

```java
ReactiveStreams.of(1).map(i -> throw new RuntimeException("failed")).onError(...)
```

could well trigger a completely different code path for handling errors in some implementations than:

```java
ReactiveStreams.failed(new Exception("failed")).onError(...)
```

Error handling before the stream has even started is likely to be different to error handling after the stream is started. So I think we should have tests for the former.

Reply:
+1, cancellation should be tested, as well as having the approaches to inject failures.

Reply:
Fixed.